lib/charms/mongodb/v0/mongodb_backups.py

```diff
@@ -0,0 +1,359 @@
+        if isinstance(pbm_status, BlockedStatus):
+            event.fail(f"Cannot restore backup {pbm_status.message}.")
+            return
+
+        # TODO restore backup
+
+    # BEGIN: helper functions
+
+    def _configure_pbm_options(self, event) -> None:
+        try:
+            self._set_config_options()
+            self._resync_config_options()
+        except SetPBMConfigError:
+            self.charm.unit.status = BlockedStatus("couldn't configure s3 backup options.")
+            return
+        except snap.SnapError as e:
+            logger.error("An exception occurred when starting pbm agent, error: %s.", str(e))
+            self.charm.unit.status = BlockedStatus("couldn't start pbm")
+            return
+        except ResyncError:
+            self.charm.unit.status = WaitingStatus("waiting to sync s3 configurations.")
+            event.defer()
+            logger.info("Deferring: Sync-ing configurations needs more time.")
+            return
+        except PBMBusyError:
+            self.charm.unit.status = WaitingStatus("waiting to sync s3 configurations.")
+            logger.info(
+                "Deferring: Cannot update configs while PBM is running, must wait for PBM action to finish."
+            )
+            event.defer()
+            return
+        except ExecError as e:
+            self.charm.unit.status = BlockedStatus(process_pbm_error(e.stdout))
+            return
+        except subprocess.CalledProcessError as e:
+            logger.error("Syncing configurations failed: %s", str(e))
+
+        self.charm.unit.status = self._get_pbm_status()
+
+    def _set_config_options(self):
+        """Applying given configurations with pbm."""
+        # TODO VM charm should implement this method
+        self.charm.set_pbm_config_file()
+
+        # the pbm tool can only set one configuration at a time.
+        for pbm_key, pbm_value in self._get_pbm_configs().items():
+            try:
+                config_cmd = ["config", "--set", f"{pbm_key}={pbm_value}"]
+                # TODO VM charm should implement this method
+                self.charm.run_pbm_command(config_cmd)
+            except (subprocess.CalledProcessError, ExecError):
+                logger.error(
+                    "Failed to configure the PBM snap option: %s",
+                    pbm_key,
+                )
+                raise SetPBMConfigError
+
+    def _get_pbm_configs(self) -> Dict:
+        """Returns a dictionary of desired PBM configurations."""
+        pbm_configs = {"storage.type": "s3"}
+        credentials = self.s3_client.get_s3_connection_info()
+        for s3_option, s3_value in credentials.items():
+            if s3_option not in S3_PBM_OPTION_MAP:
+                continue
+
+            pbm_configs[S3_PBM_OPTION_MAP[s3_option]] = s3_value
+        return pbm_configs
+
+    def _resync_config_options(self):
+        """Attempts to sync pbm config options and sets status in case of failure."""
```

Comment:
Out of curiosity, what does "syncing pbm config" mean?

Reply:
It is  the process of ensuring the PBM configuration is consistent across all nodes in MongoDB cluster.